src/mono/mono/metadata/memory-manager.c

```diff
@@ -0,0 +1,370 @@
+{
+	g_assert (memory_manager->memory_manager.is_generic);
+
+	// This should be an atomic read/write, but I'm sure it's fine
+	if (memory_manager->memory_manager.freeing)
+		return;
+
+	memory_manager->memory_manager.freeing = TRUE;
+
+	// Remove from composing ALCs
+	for (int i = 0; i < memory_manager->n_alcs; i++) {
+		MonoAssemblyLoadContext *alc = memory_manager->alcs [i];
+		mono_alc_memory_managers_lock (alc);
+		g_ptr_array_remove (alc->generic_memory_managers, memory_manager);
+		mono_alc_memory_managers_unlock (alc);
+	}
+	g_free (memory_manager->alcs);
+
+	memory_manager_delete ((MonoMemoryManager *)memory_manager, debug_unload);
+	g_free (memory_manager);
+}
+#endif
+
+gpointer
+mono_memory_manager_alloc (MonoMemoryManager *memory_manager, guint size)
+{
+	gpointer res;
+
+	mono_memory_manager_lock (memory_manager);
+#ifndef DISABLE_PERFCOUNTERS
+	mono_atomic_fetch_add_i32 (&mono_perfcounters->loader_bytes, size);
+#endif
+	res = mono_mempool_alloc (memory_manager->mp, size);
+	mono_memory_manager_unlock (memory_manager);
+
+	return res;
+}
+
+gpointer
+mono_memory_manager_alloc0 (MonoMemoryManager *memory_manager, guint size)
+{
+	gpointer res;
+
+	mono_memory_manager_lock (memory_manager);
+#ifndef DISABLE_PERFCOUNTERS
+	mono_atomic_fetch_add_i32 (&mono_perfcounters->loader_bytes, size);
+#endif
+	res = mono_mempool_alloc0 (memory_manager->mp, size);
+	mono_memory_manager_unlock (memory_manager);
+
+	return res;
+}
+
+void*
+mono_memory_manager_code_reserve (MonoMemoryManager *memory_manager, int size)
+{
+	gpointer res;
+
+	mono_memory_manager_lock (memory_manager);
+	res = mono_code_manager_reserve (memory_manager->code_mp, size);
+	mono_memory_manager_unlock (memory_manager);
+
+	return res;
+}
+
+void*
+mono_memory_manager_code_reserve_align (MonoMemoryManager *memory_manager, int size, int alignment)
+{
+	gpointer res;
+
+	mono_memory_manager_lock (memory_manager);
+	res = mono_code_manager_reserve_align (memory_manager->code_mp, size, alignment);
+	mono_memory_manager_unlock (memory_manager);
+
+	return res;
+}
+
+void
+mono_memory_manager_code_commit (MonoMemoryManager *memory_manager, void *data, int size, int newsize)
+{
+	mono_memory_manager_lock (memory_manager);
+	mono_code_manager_commit (memory_manager->code_mp, data, size, newsize);
+	mono_memory_manager_unlock (memory_manager);
+}
+
+void
+mono_memory_manager_code_foreach (MonoMemoryManager *memory_manager, MonoCodeManagerFunc func, void *user_data)
+{
+	mono_memory_manager_lock (memory_manager);
+	mono_code_manager_foreach (memory_manager->code_mp, func, user_data);
+	mono_memory_manager_unlock (memory_manager);
+}
+
+void *
+mono_method_alloc_code (MonoDomain *domain, MonoMethod *method, int size)
+{
+	MonoMemoryManager *memory_manager = mono_memory_manager_from_method (domain, method);
+	return mono_memory_manager_alloc (memory_manager, size);
+}
+
+void *
+mono_method_alloc0_code (MonoDomain *domain, MonoMethod *method, int size)
+{
```

Comment:
What does the _code suffix mean here ?

Reply:
Just that it's code memory, though maybe that's unnecessary. I wasn't sure on the nomenclature for a bunch of this.